optuna/integration/pytorch_lightning.py

```diff
@@ -56,3 +59,87 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
         if self._trial.should_prune():
             message = "Trial was pruned at epoch {}.".format(epoch)
             raise optuna.TrialPruned(message)
+
+
+class PyTorchLightningDDPPruningCallback(Callback):
+    """PyTorch Lightning DDP callback to prune unpromising trials.
+
+    See `the example <https://github.com/optuna/optuna-examples/blob/
+    main/pytorch/pytorch_lightning_simple.py>`__
+    if you want to add a pruning callback which observes accuracy.
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        monitor:
+            An evaluation metric for pruning, e.g., ``val_loss`` or
+            ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
+            ``pytorch_lightning.LightningModule.training_step`` or
+            ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
+            how this dictionary is formatted.
+    """
+
+    def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
+        _imports.check()
+
+        if not isinstance(trial.study._storage, _CachedStorage):
+            raise ValueError(
+                ":class:`~optuna.integration.PyTorchLightningPruningCallback`"
+                " supports only :class:`~optuna.storages.RDBStorage`."
+            )
+
+        super().__init__()
+
+        self._trial = trial
+        self.monitor = monitor
+
+    def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        epoch = pl_module.current_epoch
+
+        current_score = trainer.callback_metrics.get(self.monitor)
+        if current_score is None:
+            message = (
+                "The metric '{}' is not in the evaluation logs for pruning. "
+                "Please make sure you set the correct metric name.".format(self.monitor)
+            )
+            warnings.warn(message)
+            return
+
+        # When multi gpu, on_validation_end function executes multi times.
+        # So Exception is ignored to avoid IntegrityError.
+        distributed_backend = trainer.accelerator_connector.distributed_backend
+        try:
+            self._trial.report(current_score, step=epoch)
+        except sqlalchemy.exc.IntegrityError:
+            if distributed_backend is not None:
+                pass
+            else:
+                raise
+
+        if self._trial.should_prune():
+            trainer.should_stop = True
+            try:
+                self._trial.set_user_attr("pruned", True)
+                self._trial.set_user_attr("epoch", epoch)
+            except sqlalchemy.exc.IntegrityError:
+                if distributed_backend is not None:
+                    pass
+                else:
+                    raise
```

Comment:
Lisewise, 
```suggestion
        should_stop = self._trial.should_prune()
        should_stop = trainer.training_type_plugin.reduce_boolean_decision(should_stop)
        if should_stop:
            trainer.should_stop = True
            if trainer.is_global_zero:
                self._trial.set_system_attr("pruned", True)
                self._trial.set_system_attr("epoch", epoch)
```

ref https://github.com/PyTorchLightning/pytorch-lightning/blob/b294c5760eee30a995fcd400127be209b12c4d7c/pytorch_lightning/callbacks/early_stopping.py#L209

Note that user might use `user_attr` with high probability. I suggest using `system_attr` rather than `user_attr` when optuna's code stores attributes.

Reply:
note that I locally confirm that the trial is pruned.

Reply:
[meme]

`should_stop = trainer.training_type_plugin.broadcast(should_stop)` might be alternative way:

```python
        if trainer.is_global_zero:
            self._trial.report(current_score, step=epoch)
            should_stop = self._trial.should_prune()
        should_stop = trainer.training_type_plugin.broadcast(should_stop)
```